File: frontend/src/api/algorithms.ts
// src/api/algorithms.ts
import { API_BASE_URL } from "./config";

export type AlgoParam = {
  name: string;
  label: string;
  type: "int" | "float";
  min: number;
  max: number;
  step: number;
  default: number;
  description?: string | null;
};

export type AlgorithmInfo = {
  id: string;
  name: string;
  category: string;
  description: string;
  params: AlgoParam[];
};

export async function listAlgorithms(): Promise<AlgorithmInfo[]> {
  const res = await fetch(`${API_BASE_URL}/algorithms`);

  if (!res.ok) {
    throw new Error(`Failed to fetch algorithms: ${res.status}`);
  }

  return (await res.json()) as AlgorithmInfo[];
}

File: backend/app/schemas/algorithms.py
from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel


class AlgorithmParam(BaseModel):
    name: str
    label: str
    type: Literal["int", "float"]
    min: float
    max: float
    step: float
    default: float
    description: Optional[str] = None


class AlgorithmInfo(BaseModel):
    id: str
    name: str
    category: str
    description: str
    params: List[AlgorithmParam]
